File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/KDimension.java
package org.kevoree.modeling.api;

import org.kevoree.modeling.api.time.TimeTree;

import java.util.Set;

/**
 * Created by duke on 9/30/14.
 */

public interface KDimension<A extends KView, B extends KDimension, C extends KUniverse> {

    public long key();

    public void parent(Callback<B> callback);

    public void children(Callback<Set<B>> callback);

    public void fork(Callback<B> callback);

    public void save(Callback<Throwable> callback);

    public void saveUnload(Callback<Throwable> callback);

    public void delete(Callback<Throwable> callback);

    public void discard(Callback<Throwable> callback);

    public A time(Long timePoint);

    public TimeTree globalTimeTree();

    public TimeTree timeTree(String path);

    public C universe();

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/KDimension.ts
///<reference path="time/TimeTree.ts"/>

interface KDimension<A extends KView, B extends KDimension<any,any,any>, C extends KUniverse<any>> {

  key(): number;

  parent(callback: Callback<B>): void;

  children(callback: Callback<B[]>): void;

  fork(callback: Callback<B>): void;

  save(callback: Callback<Throwable>): void;

  saveUnload(callback: Callback<Throwable>): void;

  delete(callback: Callback<Throwable>): void;

  discard(callback: Callback<Throwable>): void;

  time(timePoint: number): A;

  timeTrees(keys: number[], callback: Callback<TimeTree[]>): void;

  timeTree(key: number, callback: Callback<TimeTree>): void;

  universe(): C;

}

